engine/dlib/src/dlib/file_descriptor.cpp

```diff
@@ -0,0 +1,158 @@
+// specific language governing permissions and limitations under the License.
+
+#include "file_descriptor.h"
+#include "log.h"
+#include <stdint.h>
+#include <assert.h>
+#if defined(_WIN32)
+#include <winsock2.h>
+#else
+#include <poll.h>
+#endif
+
+
+namespace dmFileDescriptor
+{
+    void PollerSetCapacity(Poller* poller, uint32_t capacity)
+    {
+        poller->m_Pollfds.SetCapacity(capacity);
+    }
+
+    int PollEventToNative(PollEvent event)
+    {
+        #if defined(_WIN32)
+        switch (event)
+        {
+            case EVENT_READ: return POLLRDNORM;
+            case EVENT_WRITE: return POLLWRNORM;
+            default:
+            case EVENT_ERROR: return POLLRDBAND;
+        }
+        #else
+        switch (event)
+        {
+            case EVENT_READ: return POLLIN;
+            // case EVENT_READ: return POLLIN | POLLPRI;
+            case EVENT_WRITE: return POLLOUT;
+            default:
+            case EVENT_ERROR: return POLLPRI;
+        }
+        #endif
+    }
+    int PollReturnEventToNative(PollEvent event)
+    {
+        #if defined(_WIN32)
+        switch (event)
+        {
+            case EVENT_READ: return POLLRDNORM;
+            case EVENT_WRITE: return POLLWRNORM;
+            default:
+            case EVENT_ERROR: return POLLHUP | POLLERR | POLLNVAL | POLLRDBAND;
+        }
+        #else
+        switch (event)
+        {
+            case EVENT_READ: return POLLIN;
+            // case EVENT_READ: return POLLIN | POLLPRI;
+            case EVENT_WRITE: return POLLOUT;
+            default:
+            case EVENT_ERROR: return POLLHUP | POLLERR | POLLNVAL;
+        }
+        #endif
+    }
+
+    void PollerClearEvent(Poller* poller, PollEvent event, int fd)
+    {
+        for (uint32_t i = 0; i < poller->m_Pollfds.Size(); ++i)
+        {
+            if (poller->m_Pollfds[i].fd == fd)
+            {
+                int e = PollEventToNative(event);
+                poller->m_Pollfds[i].events &= ~e;
+                return;
+            }
+        }
+    }
+
+    void PollerSetEvent(Poller* poller, PollEvent event, int fd)
+    {
+        int e = PollEventToNative(event);
+        for (uint32_t i = 0; i < poller->m_Pollfds.Size(); ++i)
+        {
+            if (poller->m_Pollfds[i].fd == fd)
+            {
+                poller->m_Pollfds[i].events |= e;
+                return;
+            }
+        }
+
+        if (poller->m_Pollfds.Full())
+        {
+            poller->m_Pollfds.OffsetCapacity(4);
+        }
+
+        #if defined(_WIN32)
+        WSAPOLLFD pfd;
+        pfd.fd = fd;
+        pfd.events = e;
+        poller->m_Pollfds.Push(pfd);
+        #else
+        pollfd pfd;
+        pfd.fd = fd;
+        pfd.events = e;
+        poller->m_Pollfds.Push(pfd);
+        #endif
+    }
+
+    bool PollerHasEvent(Poller* poller, PollEvent event, int fd)
+    {
+        for (uint32_t i = 0; i < poller->m_Pollfds.Size(); ++i)
```

Comment:
How large are these arrays?

Reply:
Quite often just a single file descriptor, sometimes a handful. It depends on where they're used.